walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,120 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    Collection<TimeRange> answer = new ArrayList<TimeRange>();
+
+    Collection<String> requestAttendees = new ArrayList<String>(request.getAttendees());
+    requestAttendees.addAll(request.getOptionalAttendees());
+    long requestDuration = request.getDuration();
+    
+    List<TimeRange> relevantEventTimes = filterEvents(events, requestAttendees);
+
+    System.out.println(relevantEventTimes);
+    findPossibleTimes(answer, relevantEventTimes, requestDuration);
+    
+    //If accommodating optional attendees isn't possible, only consider mandatory attendees
+    if (answer.size() == 0 && request.getAttendees().size() > 0) {
+      requestAttendees = request.getAttendees();  
+      relevantEventTimes = filterEvents(events, requestAttendees);
+      findPossibleTimes(answer, relevantEventTimes, requestDuration);
+    }
+    
```

Comment:
Per the code challenge's requirement: "Find the time slot(s) that allow mandatory attendees and the greatest possible number of optional attendees to attend". It seems your return is "timeslots that don't consider optional attendees", so they might not satisfy "the greatest possible number of optional attendees" requirement.

Reply:
Ah, yes, actually this PR isn't an attempt at the optional code challenge. This is what I implemented: "The basic functionality of optional attendees is that if one or more time slots exists so that both mandatory and optional attendees can attend, return those time slots. Otherwise, return the time slots that fit just the mandatory attendees." 

However, I will try to come back to this and try the code challenge using your tips above!